File: firmware/esp32s3/main/face_gate.cpp
#include "face_gate.h"

#include <cmath>

FaceBox face_gate_roi(const FaceBox &box, int frame_w, int frame_h) {
    FaceBox roi;
    if (!box.valid || box.w <= 0 || box.h <= 0 || frame_w <= 0 || frame_h <= 0) return roi;

    const int longest = box.w > box.h ? box.w : box.h;
    int side = static_cast<int>(static_cast<float>(longest) * (1.0f + 2.0f * FACE_ROI_PAD));
    if (side < FACE_ROI_MIN_SIDE) side = FACE_ROI_MIN_SIDE;

    const int shortest_frame = frame_w < frame_h ? frame_w : frame_h;
    // Square, so the aspect ratio the detector sees is the same whichever frame size
    // it is fed. A non-square crop would stretch the face relative to the full-frame
    // case and move the landmarks systematically.
    if (side >= static_cast<int>(static_cast<float>(shortest_frame) * FACE_ROI_MAX_FRAC)) {
        return roi;    // not worth cropping; caller uses the whole frame
    }

    // Even side and even origin. RGB565 is two bytes a pixel so any offset is
    // addressable, but keeping both even means the crop lands on the same pixel grid
    // as the full frame, which is one less way for the mapping back to drift.
    side &= ~1;

    const int cx = box.x + box.w / 2;
    const int cy = box.y + box.h / 2;
    int x = cx - side / 2;
    int y = cy - side / 2;
    if (x < 0) x = 0;
    if (y < 0) y = 0;
    if (x + side > frame_w) x = frame_w - side;
    if (y + side > frame_h) y = frame_h - side;
    x &= ~1;
    y &= ~1;

    roi.x = x;
    roi.y = y;
    roi.w = side;
    roi.h = side;
    roi.valid = true;
    return roi;
}

void face_gate_map_out(const FaceBox &roi, FaceBox *box, Landmarks *lm) {
    if (!roi.valid) return;    // detection was run on the whole frame already
    if (box != nullptr && box->valid) {
        box->x += roi.x;
        box->y += roi.y;
    }
    if (lm != nullptr && lm->valid) {
        for (int i = 0; i < 5; ++i) {
            lm->x[i] += static_cast<float>(roi.x);
            lm->y[i] += static_cast<float>(roi.y);
        }
    }
}

float face_gate_iou(const FaceBox &a, const FaceBox &b) {
    if (!a.valid || !b.valid || a.w <= 0 || a.h <= 0 || b.w <= 0 || b.h <= 0) return 0.0f;

    const int x0 = a.x > b.x ? a.x : b.x;
    const int y0 = a.y > b.y ? a.y : b.y;
    const int x1 = (a.x + a.w) < (b.x + b.w) ? (a.x + a.w) : (b.x + b.w);
    const int y1 = (a.y + a.h) < (b.y + b.h) ? (a.y + a.h) : (b.y + b.h);
    if (x1 <= x0 || y1 <= y0) return 0.0f;

    const float inter = static_cast<float>(x1 - x0) * static_cast<float>(y1 - y0);
    const float area_a = static_cast<float>(a.w) * static_cast<float>(a.h);
    const float area_b = static_cast<float>(b.w) * static_cast<float>(b.h);
    const float uni = area_a + area_b - inter;
    return uni > 0.0f ? (inter / uni) : 0.0f;
}

FaceReject face_gate_check(const Landmarks &lm, const FaceBox &box, float score,
                           int frame_w, int frame_h, FaceGeometry *measured) {
    if (measured != nullptr) *measured = FaceGeometry{};
    if (!lm.valid || !box.valid || box.w <= 0 || box.h <= 0) return FaceReject::NoLandmarks;

    // 0. Confidence, first, because it is the cheapest and because a candidate the
    //    detector itself does not believe should not be argued about geometrically.
    //    A NaN score fails this: comparisons against NaN are false either way, so it
    //    is spelled as "not >= threshold" rather than "< threshold".
    if (!(score >= FACE_MIN_SCORE)) return FaceReject::LowScore;

    // 1. Box shape and size. Frame-relative checks are skipped when the caller did
    //    not supply a frame, which is the honest thing to do - inventing a frame size
    //    would silently change what "too small" means.
    const float bw = static_cast<float>(box.w);
    const float bh = static_cast<float>(box.h);
    if (frame_w > 0 && frame_h > 0) {
        const float shortest = static_cast<float>(frame_w < frame_h ? frame_w : frame_h);
        const float side = bw > bh ? bw : bh;
        if (side < FACE_MIN_SIDE_FRAC * shortest) return FaceReject::TooSmall;
        if (side > FACE_MAX_SIDE_FRAC * shortest) return FaceReject::TooLarge;
    }
    const float aspect = bw / bh;
    if (aspect < FACE_ASPECT_MIN || aspect > FACE_ASPECT_MAX) return FaceReject::Aspect;

    // 2. The landmarks have to be near the box they came with. A detection whose
    //    keypoints land somewhere else entirely is two different objects reported as
    //    one, and every measurement taken from it would be nonsense.
    const float mx = bw * FACE_KP_MARGIN_FRAC;
    const float my = bh * FACE_KP_MARGIN_FRAC;
    const float x_lo = static_cast<float>(box.x) - mx;
    const float x_hi = static_cast<float>(box.x + box.w) + mx;
    const float y_lo = static_cast<float>(box.y) - my;
    const float y_hi = static_cast<float>(box.y + box.h) + my;
    for (int i = 0; i < 5; ++i) {
        if (lm.x[i] < x_lo || lm.x[i] > x_hi) return FaceReject::OutsideBox;
        if (lm.y[i] < y_lo || lm.y[i] > y_hi) return FaceReject::OutsideBox;
    }

    // 3. Everything else is exactly the geometry the behaviour cues are built on, so
    //    it is computed by the same function rather than re-derived here. If the
    //    geometry cannot be formed at all - degenerate eye distance - there is
    //    nothing to measure and nothing to trust.
    const FaceGeometry g = behavior_face_geometry(lm);
    if (measured != nullptr) *measured = g;
    if (!g.valid) return FaceReject::Degenerate;

    if (g.eye_dist < FACE_EYE_DIST_MIN_FRAC * bw) return FaceReject::EyeDistSmall;
    if (g.eye_dist > FACE_EYE_DIST_MAX_FRAC * bw) return FaceReject::EyeDistLarge;
    if (std::fabs(g.roll) > FACE_MAX_ROLL_DEG) return FaceReject::Roll;

    // 4. The mouth is below the eye line and the nose is between them. In the
    //    de-rotated frame that is just a sign check plus an ordering, and it is what
    //    catches a landmark set that has been permuted or fitted to something that
    //    is not a face. jaw_drop is signed, so a negative value means the detector
    //    put the mouth above the eyes.
    if (g.jaw_drop < FACE_JAW_MIN) return FaceReject::JawSmall;
    if (g.jaw_drop > FACE_JAW_MAX) return FaceReject::JawLarge;
    // nose_frac is nose height over mouth height in the same frame: 0 puts the nose
    // on the eye line, 1 puts it at the mouth. Real faces sit near 0.55.
    if (g.nose_frac < FACE_NOSE_FRAC_MIN) return FaceReject::NoseHigh;
    if (g.nose_frac > FACE_NOSE_FRAC_MAX) return FaceReject::NoseLow;

    // 5. Two lateral checks the vertical ones cannot make. Both catch the same class
    //    of failure from different directions: five points fitted to something with
    //    face-like vertical structure and no face-like horizontal structure, which is
    //    what a hand, a forearm or a seat edge produces.
    if (std::fabs(g.yaw) > FACE_YAW_MAX) return FaceReject::Yaw;
    if (g.mouth_ratio < FACE_MOUTH_MIN) return FaceReject::MouthNarrow;
    if (g.mouth_ratio > FACE_MOUTH_MAX) return FaceReject::MouthWide;

    return FaceReject::None;
}

const char *face_gate_reject_name(FaceReject r) {
    switch (r) {
        case FaceReject::None: return "ok";
        case FaceReject::NoLandmarks: return "no-landmarks";
        case FaceReject::LowScore: return "score-too-low";
        case FaceReject::TooSmall: return "face-too-small";
        case FaceReject::TooLarge: return "face-too-large";
        case FaceReject::Aspect: return "box-not-head-shaped";
        case FaceReject::OutsideBox: return "landmarks-outside-box";
        case FaceReject::Degenerate: return "degenerate-eye-distance";
        case FaceReject::EyeDistSmall: return "eye-distance-too-small";
        case FaceReject::EyeDistLarge: return "eye-distance-too-large";
        case FaceReject::Roll: return "roll-too-steep";
        case FaceReject::JawSmall: return "mouth-not-below-eyes";
        case FaceReject::JawLarge: return "mouth-too-far-below-eyes";
        case FaceReject::NoseHigh: return "nose-above-eye-line";
        case FaceReject::NoseLow: return "nose-at-or-below-mouth";
        case FaceReject::Yaw: return "nose-outside-eye-pair";
        case FaceReject::MouthNarrow: return "mouth-too-narrow";
        case FaceReject::MouthWide: return "mouth-too-wide";
        case FaceReject::Discontinuous: return "moved-too-far";
        default: return "?";
    }
}

bool face_gate_plausible(const Landmarks &lm, const FaceBox &box, float score,
                         int frame_w, int frame_h) {
    return face_gate_check(lm, box, score, frame_w, frame_h, nullptr) == FaceReject::None;
}

bool face_gate_continuous(const FaceBox &prev, const FaceBox &cand) {
    // Nothing to be discontinuous with. Answering true here is what makes the first
    // detection of a track acceptable without a special case at every call site.
    if (!prev.valid || prev.w <= 0 || prev.h <= 0) return true;
    if (!cand.valid || cand.w <= 0 || cand.h <= 0) return false;

    const float pcx = prev.x + prev.w * 0.5f;
    const float pcy = prev.y + prev.h * 0.5f;
    const float ccx = cand.x + cand.w * 0.5f;
    const float ccy = cand.y + cand.h * 0.5f;

    // Mean of the two sides, so a shrinking box and a growing one are treated the
    // same way. Using only the previous side would make a shrinking face look like it
    // had jumped further than it did.
    const float pside = 0.5f * (prev.w + prev.h);
    const float cside = 0.5f * (cand.w + cand.h);
    const float mean = 0.5f * (pside + cside);
    if (mean <= 0.0f) return false;

    const float dx = ccx - pcx;
    const float dy = ccy - pcy;
    if (std::sqrt(dx * dx + dy * dy) > FACE_JUMP_MAX_FRAC * mean) return false;

    const float ratio = pside > cside ? (pside / cside) : (cside / pside);
    return ratio <= FACE_SCALE_MAX_RATIO;
}

int face_gate_pick(const FaceBox *boxes, const Landmarks *lms, const float *scores,
                   int n, const FaceBox &track, int frame_w, int frame_h) {
    if (boxes == nullptr || lms == nullptr || n <= 0) return -1;

    int best = -1;
    float best_iou = 0.0f;
    int biggest = -1;
    long biggest_area = 0;

    for (int i = 0; i < n; ++i) {
        // A null scores array means the caller has already applied its own
        // confidence policy; 1.0 then passes the check without weakening it for
        // callers that do supply scores.
        const float score = scores != nullptr ? scores[i] : 1.0f;
        // Set FACE_GATE_ENFORCE to 0 to make this advisory - the check still runs
        // and is still logged, it just stops excluding. That is the first thing to
        // try if detection ever dies after a change here; see face_gate.h for why
        // that escape hatch exists.
        if (FACE_GATE_ENFORCE &&
            !face_gate_plausible(lms[i], boxes[i], score, frame_w, frame_h)) {
            continue;
        }

        const long area = static_cast<long>(boxes[i].w) * static_cast<long>(boxes[i].h);
        if (area > biggest_area) {
            biggest_area = area;
            biggest = i;
        }
        const float iou = face_gate_iou(boxes[i], track);
        if (iou > best_iou) {
            best_iou = iou;
            best = i;
        }
    }

    // Continuity first, size only as the opening guess. Handing over to a larger box
    // the moment one appears is how a passenger takes the driver's alarm.
    if (best >= 0 && best_iou >= FACE_TRACK_MIN_IOU) return best;
    return biggest;
}

// ---------------- FaceTrack ----------------

FaceTrackResult FaceTrack::snapshot() const {
    FaceTrackResult r;
    r.present = present();
    r.confirmed = confirmed_;
    r.misses = misses_;
    r.agreed = agreed_;
    r.box = box_;
    r.lm = lm_;
    return r;
}

FaceTrackResult FaceTrack::peek() const { return snapshot(); }

FaceTrackResult FaceTrack::update(bool have, const FaceBox &box, const Landmarks &lm,
                                  float score, int frame_w, int frame_h) {
    const bool was_present = present();

    FaceReject why = FaceReject::None;
    bool accept = false;

    // Whether this candidate could be the same face as the tracked one. Computed
    // unconditionally, because it answers a question about the world; how much
    // weight to give it is a separate decision, taken twice below.
    bool same_face = false;

    if (have) {
        why = face_gate_check(lm, box, score, frame_w, frame_h, nullptr);
        if (why == FaceReject::None) {
            same_face = face_gate_continuous(box_, box);
            // While the track is warm, a candidate somewhere else is a different
            // object and is refused outright - that is what stops a passenger taking
            // over a live track. Once the face has been missing for
            // FACE_REACQUIRE_AFTER attempts the driver may genuinely have moved, so a
            // discontinuous candidate is allowed in; see below for what it costs.
            const bool warm = box_.valid && misses_ < FACE_REACQUIRE_AFTER;
            if (warm && !same_face) {
                why = FaceReject::Discontinuous;
            } else {
                accept = true;
            }
        }
    }

    if (accept) {
        // A candidate that does not line up with what we had starts a NEW track
        // rather than inheriting the old one's confirmation. Presence is re-earned
        // across a discontinuity, never inherited - that is the property that stops
        // a track sliding off the driver and onto whatever is behind them.
        //
        // Note what this deliberately does not do: a face that reappears where it
        // was, after any number of misses inside the hold, DOES continue the track.
        // Ending the track there would drop presence for a detection interval every
        // time the detector blinked, which is exactly the moment - eyes closing, head
        // nodding - that the hold exists to carry.
        if (box_.valid && !same_face) {
            agreed_ = 0;
            confirmed_ = false;
        }
        if (agreed_ < FACE_CONFIRM_DETECTIONS) ++agreed_;
        box_ = box;
        lm_ = lm;
        misses_ = 0;
        if (agreed_ >= FACE_CONFIRM_DETECTIONS) confirmed_ = true;
    } else {
        ++misses_;
        // A pending track has nothing to hold: it was never believed in the first
        // place, so one miss ends it rather than leaving a half-formed track around
        // to be resumed by an unrelated candidate.
        if (!confirmed_) {
            agreed_ = 0;
            box_.valid = false;
            lm_.valid = false;
        } else if (misses_ > FACE_HOLD_DETECTIONS) {
            // The hold has expired. Drop everything, so the next detection is a fresh
            // acquisition and the caller can tell the detector to stop searching where
            // the driver used to be.
            confirmed_ = false;
            agreed_ = 0;
            box_.valid = false;
            lm_.valid = false;
        }
    }

    FaceTrackResult r = snapshot();
    r.fresh = accept;
    r.reject = accept ? FaceReject::None : why;
    r.acquired = r.present && !was_present;
    r.lost = was_present && !r.present;
    return r;
}

void FaceTrack::reset() {
    box_ = FaceBox{};
    lm_ = Landmarks{};
    agreed_ = 0;
    misses_ = 0;
    confirmed_ = false;
}

File: src/drowsyguard/facegate.py
"""Desktop mirror of ``firmware/esp32s3/main/face_gate.{h,cpp}``.

Which detection to trust, and when to believe a driver is really there. The firmware
file is the reference implementation and carries the full reasoning for every
constant and every check; this module exists so the dashboard applies *the same*
decisions, because thresholds tuned on a laptop and enforced only on a laptop are
worth nothing.

``tests/test_facegate_parity.py`` compiles the C++ and drives both through the same
sequences, asserting the same verdict, the same reject reason and the same presence
decision at every step. That is a stronger guarantee than matching constants, and it
is the reason this file is a transcription rather than a reinterpretation: where a
Python idiom would read better but behave even slightly differently at a boundary,
the C++ shape wins.

Two kinds of check, and keeping them apart is the design:

* :func:`check` is *static* - one candidate, on its own. Confidence, box size and
  shape, then the five landmarks: eyes level-ish, mouth below them, nose between
  them and inside the eye pair, mouth narrower than the head. It is cheap, it is
  order-dependent (the first failure is the one reported), and it cannot see time.

* :class:`FaceTrack` is *temporal*. It requires consecutive detections to agree
  before calling a driver present, refuses a candidate that has moved further than a
  head can move between detections, holds the last box briefly when the detector
  misses, and - the part that is easy to get wrong - never lets a new track inherit
  the old one's confirmation. The last property is what stops a track sliding off
  the driver and onto a passenger.
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from enum import Enum

from .behavior import FaceGeometry, face_geometry

# --- confidence -------------------------------------------------------------
FACE_MIN_SCORE = 0.55

# --- plausibility limits ----------------------------------------------------
FACE_EYE_DIST_MIN_FRAC = 0.15
FACE_EYE_DIST_MAX_FRAC = 0.95
FACE_MAX_ROLL_DEG = 45.0
FACE_JAW_MIN = 0.35
FACE_JAW_MAX = 2.60
FACE_NOSE_FRAC_MIN = 0.15
FACE_NOSE_FRAC_MAX = 1.15
FACE_YAW_MAX = 0.75
FACE_MOUTH_MIN = 0.25
FACE_MOUTH_MAX = 1.60

# --- box size and shape -----------------------------------------------------
FACE_MIN_SIDE_FRAC = 0.10
FACE_MAX_SIDE_FRAC = 0.95
FACE_ASPECT_MIN = 0.55
FACE_ASPECT_MAX = 1.80
FACE_KP_MARGIN_FRAC = 0.30

# --- temporal consistency ---------------------------------------------------
FACE_CONFIRM_DETECTIONS = 2
FACE_HOLD_DETECTIONS = 5
FACE_JUMP_MAX_FRAC = 1.20
FACE_SCALE_MAX_RATIO = 1.80
FACE_REACQUIRE_AFTER = 2

# --- region of interest -----------------------------------------------------
FACE_TRACK_MIN_IOU = 0.15


class FaceReject(Enum):
    """Why a candidate was not believed.

    The string values are the firmware's ``face_gate_reject_name()`` output verbatim,
    so a rejection reason read off the device's status page and one printed by the
    dashboard are the same token and can be grepped together.
    """

    NONE = 'ok'
    NO_LANDMARKS = 'no-landmarks'
    LOW_SCORE = 'score-too-low'
    TOO_SMALL = 'face-too-small'
    TOO_LARGE = 'face-too-large'
    ASPECT = 'box-not-head-shaped'
    OUTSIDE_BOX = 'landmarks-outside-box'
    DEGENERATE = 'degenerate-eye-distance'
    EYE_DIST_SMALL = 'eye-distance-too-small'
    EYE_DIST_LARGE = 'eye-distance-too-large'
    ROLL = 'roll-too-steep'
    JAW_SMALL = 'mouth-not-below-eyes'
    JAW_LARGE = 'mouth-too-far-below-eyes'
    NOSE_HIGH = 'nose-above-eye-line'
    NOSE_LOW = 'nose-at-or-below-mouth'
    YAW = 'nose-outside-eye-pair'
    MOUTH_NARROW = 'mouth-too-narrow'
    MOUTH_WIDE = 'mouth-too-wide'
    DISCONTINUOUS = 'moved-too-far'


@dataclass(frozen=True)
class Box:
    """A face box in frame pixels, as (x, y, w, h). Not square; that is the point.

    ``FaceTracker`` downstream turns the accepted box into a square crop, but the
    gate has to see the detector's own rectangle - squaring it first would destroy
    the aspect-ratio check, which is one of the two things that reject a hand.
    """

    x: float = 0.0
    y: float = 0.0
    w: float = 0.0
    h: float = 0.0
    valid: bool = False

    @property
    def cx(self):
        return self.x + self.w / 2.0

    @property
    def cy(self):
        return self.y + self.h / 2.0

    @property
    def side(self):
        """Mean of the two sides. What continuity is measured in units of."""
        return 0.5 * (self.w + self.h)

    def as_square(self):
        """(x0, y0, side) square crop, the shape the eye-patch code expects."""
        side = max(self.w, self.h)
        return (int(round(self.cx - side / 2.0)),
                int(round(self.cy - side / 2.0)),
                int(round(side)))


def iou(a: Box, b: Box) -> float:
    if not a.valid or not b.valid or a.w <= 0 or a.h <= 0 or b.w <= 0 or b.h <= 0:
        return 0.0
    x0, y0 = max(a.x, b.x), max(a.y, b.y)
    x1, y1 = min(a.x + a.w, b.x + b.w), min(a.y + a.h, b.y + b.h)
    if x1 <= x0 or y1 <= y0:
        return 0.0
    inter = (x1 - x0) * (y1 - y0)
    union = a.w * a.h + b.w * b.h - inter
    return inter / union if union > 0 else 0.0


def check(landmarks, box: Box, score: float, frame_w=0, frame_h=0):
    """Static plausibility. Returns ``(FaceReject, FaceGeometry)``.

    The geometry is returned alongside the verdict rather than recomputed by the
    caller, because a rejection you cannot see the numbers behind is not actionable -
    that lesson cost this project a debugging session in which the gate rejected
    100% of real faces and reported only a count.

    ``frame_w``/``frame_h`` of 0 skip the two frame-relative size checks, which is
    the honest response to a caller that has no frame rather than inventing one.
    """
    empty = FaceGeometry()
    if not landmarks or len(landmarks) < 5 or not box.valid or box.w <= 0 or box.h <= 0:
        return FaceReject.NO_LANDMARKS, empty

    # NaN-safe by construction: `not (score >= t)` is true for NaN, where
    # `score < t` would be false and let it through.
    if not (score >= FACE_MIN_SCORE):
        return FaceReject.LOW_SCORE, empty

    if frame_w > 0 and frame_h > 0:
        shortest = float(min(frame_w, frame_h))
        side = max(box.w, box.h)
        if side < FACE_MIN_SIDE_FRAC * shortest:
            return FaceReject.TOO_SMALL, empty
        if side > FACE_MAX_SIDE_FRAC * shortest:
            return FaceReject.TOO_LARGE, empty

    aspect = box.w / box.h
    if aspect < FACE_ASPECT_MIN or aspect > FACE_ASPECT_MAX:
        return FaceReject.ASPECT, empty

    mx, my = box.w * FACE_KP_MARGIN_FRAC, box.h * FACE_KP_MARGIN_FRAC
    x_lo, x_hi = box.x - mx, box.x + box.w + mx
    y_lo, y_hi = box.y - my, box.y + box.h + my
    for px, py in landmarks[:5]:
        if px < x_lo or px > x_hi or py < y_lo or py > y_hi:
            return FaceReject.OUTSIDE_BOX, empty

    g = face_geometry(landmarks)
    if not g.valid:
        return FaceReject.DEGENERATE, g

    if g.eye_dist < FACE_EYE_DIST_MIN_FRAC * box.w:
        return FaceReject.EYE_DIST_SMALL, g
    if g.eye_dist > FACE_EYE_DIST_MAX_FRAC * box.w:
        return FaceReject.EYE_DIST_LARGE, g
    if abs(g.roll) > FACE_MAX_ROLL_DEG:
        return FaceReject.ROLL, g
    if g.jaw_drop < FACE_JAW_MIN:
        return FaceReject.JAW_SMALL, g
    if g.jaw_drop > FACE_JAW_MAX:
        return FaceReject.JAW_LARGE, g
    if g.nose_frac < FACE_NOSE_FRAC_MIN:
        return FaceReject.NOSE_HIGH, g
    if g.nose_frac > FACE_NOSE_FRAC_MAX:
        return FaceReject.NOSE_LOW, g
    if abs(g.yaw) > FACE_YAW_MAX:
        return FaceReject.YAW, g
    if g.mouth_ratio < FACE_MOUTH_MIN:
        return FaceReject.MOUTH_NARROW, g
    if g.mouth_ratio > FACE_MOUTH_MAX:
        return FaceReject.MOUTH_WIDE, g
    return FaceReject.NONE, g


def plausible(landmarks, box: Box, score: float, frame_w=0, frame_h=0) -> bool:
    return check(landmarks, box, score, frame_w, frame_h)[0] is FaceReject.NONE


def continuous(prev: Box, cand: Box) -> bool:
    """Could `cand` be the same face as `prev`, one detection interval later?

    Scale-free in both terms, so it behaves identically for a driver sitting close
    and one sitting back. An invalid `prev` answers True - there is nothing to be
    discontinuous with, and answering False would need a special case at every call
    site to let a first detection through.
    """
    if not prev.valid or prev.w <= 0 or prev.h <= 0:
        return True
    if not cand.valid or cand.w <= 0 or cand.h <= 0:
        return False
    mean = 0.5 * (prev.side + cand.side)
    if mean <= 0:
        return False
    if math.hypot(cand.cx - prev.cx, cand.cy - prev.cy) > FACE_JUMP_MAX_FRAC * mean:
        return False
    ratio = (prev.side / cand.side) if prev.side > cand.side else (cand.side / prev.side)
    return ratio <= FACE_SCALE_MAX_RATIO


def pick(candidates, track: Box, frame_w=0, frame_h=0) -> int:
    """Which candidate is the driver. `candidates` is [(Box, landmarks, score), ...].

    Continuity first, size only as the opening guess: with a live track the right
    answer is the box that overlaps it, because a passenger leaning forward is a
    bigger face than the driver and "largest wins" hands them the alarm. Returns the
    index, or -1.
    """
    best, best_iou = -1, 0.0
    biggest, biggest_area = -1, 0.0
    for i, (box, lms, score) in enumerate(candidates):
        if not plausible(lms, box, score, frame_w, frame_h):
            continue
        area = box.w * box.h
        if area > biggest_area:
            biggest_area, biggest = area, i
        overlap = iou(box, track)
        if overlap > best_iou:
            best_iou, best = overlap, i
    if best >= 0 and best_iou >= FACE_TRACK_MIN_IOU:
        return best
    return biggest


@dataclass
class TrackResult:
    present: bool = False
    fresh: bool = False
    confirmed: bool = False
    acquired: bool = False
    lost: bool = False
    reject: FaceReject = FaceReject.NONE
    misses: int = 0
    agreed: int = 0
    box: Box = field(default_factory=Box)
    landmarks: list = field(default_factory=list)


class FaceTrack:
    """The driver, across time. Mirror of ``FaceTrack`` in face_gate.h."""

    def __init__(self):
        self.reset()

    def reset(self):
        self._box = Box()
        self._lm = []
        self._agreed = 0
        self._misses = 0
        self._confirmed = False

    @property
    def present(self) -> bool:
        return self._confirmed and self._misses <= FACE_HOLD_DETECTIONS

    @property
    def box(self) -> Box:
        return self._box

    @property
    def landmarks(self):
        return list(self._lm)

    def _snapshot(self) -> TrackResult:
        return TrackResult(present=self.present, confirmed=self._confirmed,
                           misses=self._misses, agreed=self._agreed,
                           box=self._box, landmarks=list(self._lm))

    def peek(self) -> TrackResult:
        """State without offering a detection.

        Frames between detection attempts still pass, but they are not evidence
        either way, so this never expires the hold: the hold is counted in attempts,
        which keeps it independent of how often the caller chooses to detect.
        """
        return self._snapshot()

    def update(self, have, box: Box, landmarks, score, frame_w=0, frame_h=0) -> TrackResult:
        """One detection *attempt*. `have` is False when the detector found nothing."""
        was_present = self.present

        why = FaceReject.NONE
        accept = False
        # Whether this candidate could be the same face as the tracked one. Computed
        # unconditionally, because it answers a question about the world; how much
        # weight to give it is a separate decision, taken twice below.
        same_face = False
        if have:
            why = check(landmarks, box, score, frame_w, frame_h)[0]
            if why is FaceReject.NONE:
                same_face = continuous(self._box, box)
                # While the track is warm, a candidate somewhere else is a different
                # object and is refused outright - that is what stops a passenger
                # taking over a live track. After FACE_REACQUIRE_AFTER misses the
                # driver may genuinely have moved, so it is allowed in.
                warm = self._box.valid and self._misses < FACE_REACQUIRE_AFTER
                if warm and not same_face:
                    why = FaceReject.DISCONTINUOUS
                else:
                    accept = True

        if accept:
            # A candidate that does not line up with what we had starts a NEW track
            # rather than inheriting the old one's confirmation. Presence is re-earned
            # across a discontinuity, never inherited.
            #
            # A face that reappears where it was, after any number of misses inside
            # the hold, DOES continue the track: ending it there would drop presence
            # every time the detector blinked, which is the moment the hold exists for.
            if self._box.valid and not same_face:
                self._agreed = 0
                self._confirmed = False
            if self._agreed < FACE_CONFIRM_DETECTIONS:
                self._agreed += 1
            self._box = box
            self._lm = list(landmarks[:5])
            self._misses = 0
            if self._agreed >= FACE_CONFIRM_DETECTIONS:
                self._confirmed = True
        else:
            self._misses += 1
            if not self._confirmed:
                # A pending track has nothing to hold: it was never believed, so one
                # miss ends it rather than leaving a half-formed track for an
                # unrelated candidate to resume.
                self._agreed = 0
                self._box = Box()
                self._lm = []
            elif self._misses > FACE_HOLD_DETECTIONS:
                self._confirmed = False
                self._agreed = 0
                self._box = Box()
                self._lm = []

        out = self._snapshot()
        out.fresh = accept
        out.reject = FaceReject.NONE if accept else why
        out.acquired = out.present and not was_present
        out.lost = was_present and not out.present
        return out
